File: pkg/module/free_page_about.go
package module

import (
	_ "embed" // embed for free_page_about.c
	"github.com/xcphoenix/elf-load-analyser/pkg/data"
	"github.com/xcphoenix/elf-load-analyser/pkg/data/form"
	"github.com/xcphoenix/elf-load-analyser/pkg/ebpf"
	"github.com/xcphoenix/elf-load-analyser/pkg/helper"
	"github.com/xcphoenix/elf-load-analyser/pkg/monitor"
	"github.com/xcphoenix/elf-load-analyser/pkg/render/enhance"
)

//go:embed src/free_page_about.c.k
var tlbAboutSource string

type tlbAboutEvent struct {
	enhance.TimeEventResult
	Type uint32
}

func (t tlbAboutEvent) Render() *data.AnalyseData {
	return data.NewAnalyseData(form.NewMarkdown(
		helper.IfElse(t.Type == 1, "为了页表的清除初始化 mmu_gather 结构体", "结束 mmu_gather 结构体").(string)),
	)
}

type freePgdRangeEventType struct {
	enhance.TimeEventResult
	Addr    uint64
	End     uint64
	Floor   uint64
	Ceiling uint64
}

func (f freePgdRangeEventType) Render() *data.AnalyseData {
	res := data.NewSet(
		form.NewMarkdown("释放可以清除的页表"),
		form.NewFmtList(form.Fmt{
			{"addr: %x, end: %x", f.Addr, f.End},
			{"floor: %x, ceiling: %x", f.Floor, f.Ceiling},
		}),
	)
	return data.NewAnalyseData(res)
}

func init() {
	m := monitor.NewPerfMonitor(&monitor.Monitor{
		Name:   "mmu_gather",
		Source: tlbAboutSource,
		Events: []*ebpf.Event{
			ebpf.NewKprobeEvent("kprobe__tlb_gather_mmu", "tlb_gather_mmu", -1),
			ebpf.NewKprobeEvent("kprobe__tlb_finish_mmu", "tlb_finish_mmu", -1),
			ebpf.NewKprobeEvent("kprobe__shift_arg_pages", "shift_arg_pages", -1),
			ebpf.NewKprobeEvent("kprobe__free_pgd_range", "free_pgd_range", -1),
		},
	})
	m.RegisterOnceTable("tlb_gather_mmu_events", monitor.RenderHandler(tlbAboutEvent{}, nil))
	m.RegisterOnceTable("tlb_finish_mmu_events", monitor.RenderHandler(tlbAboutEvent{}, nil))
	m.RegisterOnceTable("free_pgd_range_events", monitor.RenderHandler(freePgdRangeEventType{}, nil))
	//factory.Register(m)
}

File: pkg/module/src/free_page_about.c
#include "common.h"

TDATA(free_page_about_event_type, u32 type);
TDATA(free_pgd_range_event_type,  // free_pgd_range
      u64 addr;                   // addr
      u64 end;                    // end
      u64 floor;                  // floor
      u64 ceiling;                // ceiling
)

BPF_PERF_OUTPUT(tlb_gather_mmu_events);
BPF_PERF_OUTPUT(tlb_finish_mmu_events);
BPF_PERF_OUTPUT(free_pgd_range_events);

BPF_PERCPU_ARRAY(total_array, u32, 1);

int kprobe__shift_arg_pages(struct pt_regs *ctx) {
    if ((u32)bpf_get_current_pid_tgid() != _PID_) {
        return 0;
    }

    int zero = 0, one = 1;
    total_array.update(&zero, &one);

    return 0;
}

int kprobe__tlb_gather_mmu(struct pt_regs *ctx) {
    if ((u32)bpf_get_current_pid_tgid() != _PID_) {
        return 0;
    }

    int zero = 0;
    u32 *cnt = total_array.lookup(&zero);
    if (cnt == NULL || *cnt != 1) {
        return 0;
    }

    struct free_page_about_event_type e = {.type = 1};
    init_tdata(&e);
    tlb_gather_mmu_events.perf_submit((void *)ctx, (void *)&e, sizeof(e));

    return 0;
}

int kprobe__tlb_finish_mmu(struct pt_regs *ctx) {
    if ((u32)bpf_get_current_pid_tgid() != _PID_) {
        return 0;
    }

    int zero = 0;
    u32 *cnt = total_array.lookup(&zero);
    if (cnt == NULL || *cnt != 1) {
        return 0;
    }

    struct free_page_about_event_type e = {.type = 2};
    init_tdata(&e);
    tlb_finish_mmu_events.perf_submit((void *)ctx, (void *)&e, sizeof(e));

    return 0;
}

int kprobe__free_pgd_range(struct pt_regs *ctx, struct mmu_gather *tlb,
                           unsigned long addr, unsigned long end,
                           unsigned long floor, unsigned long ceiling) {
    if ((u32)bpf_get_current_pid_tgid() != _PID_) {
        return 0;
    }

    int zero = 0;
    u32 *cnt = total_array.lookup(&zero);
    if (cnt == NULL || *cnt != 1) {
        return 0;
    }

    struct free_pgd_range_event_type e = {};
    init_tdata(&e);
    e.addr    = (u64)addr;
    e.end     = (u64)end;
    e.floor   = (u64)floor;
    e.ceiling = (u64)ceiling;

    free_pgd_range_events.perf_submit((void*)ctx, (void*)&e, sizeof(e));

    return 0;
}